### PR TITLE
feat: Add ui-react-liveness and FaceLivenessDetector component

### PR DESCRIPTION
**Pull request by thaddmt**
#### Description of changes
- This change introduces the `ui-react-liveness` package and the `FaceLivenessDetector` component
- Includes documentation changes
- NOTE: make sure to squash and merge when merging this change
- Note to @aws-amplify/amplify-ui team members, please pay extra attention to files outside of the `react-liveness` package to ensure that we are not changing the prod GH actions and tests. Particularly the `.github` folder

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are updated
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/changeset-presets/bump-react.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 '@aws-amplify/ui-react': minor
 '@aws-amplify/ui-react-core': patch
 '@aws-amplify/ui-react-native': major
+'@aws-amplify/ui-react-liveness': minor
```
Comment (ioanabrooks) [inline .github/changeset-presets/bump-react.md:6]: Can we consolidate the changeset into a single one? Also, don't think `ui-react-native` needs a major bump.

File: packages/react-native/src/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export function capitalize<T extends string>(value: T): Capitalize<T> {
```
Comment (ioanabrooks) [inline packages/react-native/src/utils/utils.ts:1]: This util was moved to the ui package.

File: .github/changeset-presets/bump-react.md (hunk 1 above)
Comment (thaddmt) [reply .github/changeset-presets/bump-react.md:6]: Hm do you mean combining `bump-react.md` and `bump-versions.md`? If so I'd prefer not to do that in this PR since this is just appending to our existing verion bumps

File: .github/actions/codeartifact-action/action.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+name: 'Return AWS CodeArtifact configuration'
```
Comment (ioanabrooks) [inline .github/actions/codeartifact-action/action.yml:1]: Is this action in use anywhere?

File: canary/apps/react/next/package.json [hunk 4]
```diff
@@ -9,7 +9,8 @@
     "lint": "next lint"
   },
   "dependencies": {
-    "@aws-amplify/ui-react": "latest",
+    "@aws-amplify/ui-react": "liveness",
+    "@aws-amplify/ui-react-liveness": "liveness",
```
Comment (ioanabrooks) [inline canary/apps/react/next/package.json:13]: Does this look right? :)

File: .github/actions/codeartifact-action/action.yml (hunk 3 above)
Comment (thaddmt) [reply .github/actions/codeartifact-action/action.yml]: Just removed that action as well!

File: canary/e2e/features/liveness/face-detect.feature [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Feature: Liveness Start Screen
+
+  Test Liveness CDN
+
+  Background:
+    Given I'm running the example "/"
```
Comment (ioanabrooks) [inline canary/e2e/features/liveness/face-detect.feature:6]: Let's write an example specifically for liveness. Seems we are re-using [overloading] the sign-in example here?

File: docs/package.json [hunk 6]
```diff
@@ -31,7 +31,7 @@
     "@mdx-js/react": "^2.1.0",
     "@moxy/next-compile-node-modules": "^2.1.1",
     "@xstate/inspect": "^0.4.1",
-    "aws-amplify": "latest",
+    "aws-amplify": "^5.0.1",
```
Comment (ioanabrooks) [inline docs/package.json:34]: Should be latest, right?

File: canary/e2e/features/liveness/face-detect.feature (hunk 5 above)
Comment (thaddmt) [reply canary/e2e/features/liveness/face-detect.feature:6]: I think the goal here was to ensure that our library can build no matter which portions of the app you are using, @0618 maybe you can chime in here

File: docs/src/data/frameworks.ts [hunk 7]
```diff
@@ -27,9 +33,11 @@ export const REACT_NATIVE_DEPENDENCIES =
...
-  vue: '@aws-amplify/ui-vue aws-amplify',
-  angular: '@aws-amplify/ui-angular aws-amplify',
-  flutter: 'amplify_authenticator',
-  'react-native': `@aws-amplify/ui-react-native aws-amplify ${REACT_NATIVE_DEPENDENCIES}`,
+  react: 'npm i @aws-amplify/ui-react aws-amplify',
+  vue: 'npm i @aws-amplify/ui-vue aws-amplify',
+  angular: 'npm i @aws-amplify/ui-angular aws-amplify',
+  flutter: 'flutter pub add amplify_authenticator',
+  android: "implementation 'com.amplifyframework.ui:liveness:1.0.0'",
+  swift: 'add swift install script here',
```
Comment (ioanabrooks) [inline docs/src/data/frameworks.ts:41]: When will this be added?

Comment (thaddmt) [reply docs/src/data/frameworks.ts:41]: This was missed, i'll sync with @atierian

File: docs/src/data/test/props-table.test.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import * as fs from 'fs';
+
+test('Props Table', () => {
```
Comment (ioanabrooks) [inline docs/src/data/test/props-table.test.ts:3]: This file shouldn't be added: https://github.com/aws-amplify/amplify-ui/blob/main/docs/tests/props-table.test.ts

File: canary/apps/react/next/package.json (hunk 4 above)
Comment (thaddmt) [reply canary/apps/react/next/package.json:13]: I think we can't change this until the package actually releases since it is installing from npm @0618 thoughts?

File: .github/changeset-presets/bump-react.md (hunk 1 above)
Comment (ioanabrooks) [reply .github/changeset-presets/bump-react.md:6]: Concerned about `'@aws-amplify/ui-react-native': major`  - don't think that's right

File: examples/react-native/src/assets/icons/index.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { ImageSourcePropType } from 'react-native';
```
Comment (ioanabrooks) [inline examples/react-native/src/assets/icons/index.ts:1]: Think all the changes under `examples/react-native` can be removed from this PR

File: docs/src/data/frameworks.ts (hunk 7 above)
Comment (atierian) [reply docs/src/data/frameworks.ts:41]: Thanks for catching this. 
There isn't an install script for Swift. The package is added through the IDE (Xcode). I don't know enough about the docs structure to say whether `swift:` should be removed or if it should just be an empty string though.

File: canary/e2e/features/liveness/face-detect.feature (hunk 5 above)
Comment (ioanabrooks) [reply canary/e2e/features/liveness/face-detect.feature:6]: Think we should use a different example to run, like `ui/components/liveness/cdn` or something example `/` is used by authenticator. might be confusing to see sign-in and sing-up tests fail bc os liveness issue.

File: canary/apps/react/next/package.json (hunk 4 above)
Comment (0618) [reply canary/apps/react/next/package.json:13]: Yes, `liveness` is the version in CA

File: .github/changeset-presets/bump-react.md (hunk 1 above)
Comment (thaddmt) [reply .github/changeset-presets/bump-react.md:6]: Think that is currently what we have in `ui/main` though right?

File: canary/e2e/features/liveness/face-detect.feature (hunk 5 above)
Comment (0618) [reply canary/e2e/features/liveness/face-detect.feature:6]: @ioanabrooks does the path matter? If there's a cdn issue, face-detect would fail, but sign-in and sign-up should be fine, right?

File: .github/changeset-presets/bump-react.md (hunk 1 above)
Comment (ioanabrooks) [reply .github/changeset-presets/bump-react.md:6]: My bad, this is the presets not an actual changeset. Ignore my comments. That being said...do we need a changeset for this PR?

Comment (thaddmt) [reply .github/changeset-presets/bump-react.md:6]: we probably do yep!

File: .github/workflows/reusable-unit.yml [hunk 10]
```diff
@@ -28,6 +28,7 @@ jobs:
           - react-native
           - react-core
           - react-storage
+          - react-liveness
```
Comment (ioanabrooks) [inline .github/workflows/reusable-unit.yml:31]: Thinking you also need to build any dependencies, not just add it to the matrix? DO you need to build react package perhaps? See: https://github.com/aws-amplify/amplify-ui/blob/main/.github/workflows/reusable-unit.yml#L83

File: docs/src/data/frameworks.ts (hunk 7 above)
Comment (thaddmt) [reply docs/src/data/frameworks.ts]: Removed the install script for swift for now

File: canary/e2e/features/liveness/face-detect.feature (hunk 5 above)
Comment (ioanabrooks) [reply canary/e2e/features/liveness/face-detect.feature:6]: Does the page render okay in case of failure?

File: .github/workflows/reusable-unit.yml (hunk 10 above)
Comment (thaddmt) [reply .github/workflows/reusable-unit.yml:31]: yep, good callout

File: packages/react-core/src/Authenticator/hooks/__mock__/components.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
+import { Defaults, Overrides } from '../types';
```
Comment (calebpollman) [inline packages/react-core/src/Authenticator/hooks/__mock__/components.ts:1]: Please remove this file 😄

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/constants.ts [hunk 12]
```diff
@@ -1,2 +1,16 @@
+import { AuthenticatorRouteComponentKey } from './types';
+
```
Comment (calebpollman) [inline packages/react-core/src/Authenticator/hooks/useAuthenticator/constants.ts:2]:
```suggestion
```

File: packages/react-core/src/Authenticator/hooks/useAuthenticator/constants.ts [hunk 13]
```diff
@@ -1,2 +1,16 @@
...
+  'signIn',
+  'signUp',
+  'forceNewPassword',
+  'confirmResetPassword',
+  'confirmSignIn',
+  'confirmSignUp',
+  'confirmVerifyUser',
+  'resetPassword',
+  'setupTOTP',
+];
```
Comment (calebpollman) [inline packages/react-core/src/Authenticator/hooks/useAuthenticator/constants.ts:16]:
```suggestion
```

File: packages/react-liveness/jest.config.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  collectCoverage: true,
+  collectCoverageFrom: ['<rootDir>/src/**/*.(ts|tsx)'],
+  // ignore coverage for top level "export", version and style files
+  coveragePathIgnorePatterns: ['<rootDir>/src/(index|styles|version).(ts|tsx)'],
+  coverageThreshold: {
+    global: {
+      branches: 65,
+      functions: 77,
+      lines: 77,
+      statements: 78,
```
Comment (calebpollman) [inline packages/react-liveness/jest.config.js:11]: Can we add a follow up task to get these all to a minimum of 80?

File: yarn.lock [hunk 15]
```diff
@@ -465,6 +465,37 @@
     tslib "^1.8.0"
     uuid "^3.2.1"
 
+"@aws-amplify/analytics@6.0.24":
```
Comment (calebpollman) [inline yarn.lock:468]: Think these yarn.lock updates are extraneous, can we check out the _yarn.lock_ from `main` and rerun `yarn` to prevent these?

**Review by calebpollman — COMMENTED** [review]
Left some feedback but overall looking good!

File: yarn.lock (hunk 15 above)
Comment (thaddmt) [reply yarn.lock:468]: doing that right now

File: packages/react-liveness/jest.config.js (hunk 14 above)
Comment (thaddmt) [reply packages/react-liveness/jest.config.js:11]: https://app.asana.com/0/1202067976849710/1204357382742891

File: .changeset/brown-moles-grow.md [hunk 16]
```diff
@@ -0,0 +1,5 @@
+---
+"docs": patch
```
Comment (calebpollman) [inline .changeset/brown-moles-grow.md:2]:
```suggestion
"@aws-amplify/ui": minor
"@aws-amplify/ui-react": minor
"@aws-amplify/ui-react-liveness": major
```

File: .github/workflows/test-internal-prs.yml [hunk 17]
```diff
@@ -111,7 +111,7 @@ jobs:
       repository: ${{ github.repository }}
 
   e2e:
-    uses: aws-amplify/amplify-ui/.github/workflows/reusable-e2e.yml@main
+    uses: aws-amplify/amplify-ui/.github/workflows/reusable-e2e.yml@liveness-main
```
Comment (wlee221) [inline .github/workflows/test-internal-prs.yml:114]: We should change to `@main` eventually before merge. This means that we have to test e2e workflow in some other PR, but looks like @0618  already opened one: #3678.

File: canary/apps/react/next/package.json (hunk 4 above)
Comment (wlee221) [reply canary/apps/react/next/package.json:13]: This means that `build-and-runtime` test will fail after this is merged onto `main`. I believe we should skip this canary change until liveness is released, or we can do a tagged release to `@liveness`.

File: environments/liveness/README.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+   ```shell
+   isengardcli assume livenesssampleserverapp-app --role Admin --region us-east-1
+   ```
+   This will configure a local AWS profile which uses temporary access keys from our shared `livenesssampleserverapp-app` account that can be used for accessing our shared environments.
+1. From Isengard, navigate to 413330344011 Admin console or using [https://tiny.amazon.com/1k31d1u4c/IsenLink](https://tiny.amazon.com/1k31d1u4c/IsenLink)
+1. Run the following command to pull the credentials
+   ```shell
+   yarn pull
+   ```
+1. Accept all the default settings
```
Comment (wlee221) [inline environments/liveness/README.md:33]: Let's remove these instructions

File: .github/workflows/reusable-e2e.yml [hunk 19]
```diff
@@ -228,6 +234,22 @@ jobs:
       - name: Pull down Datastore AWS environments
         run: yarn environments datastore pull
 
+      - name: Configure liveness credentials
+        uses: aws-actions/configure-aws-credentials@v1
```
Comment (wlee221) [inline .github/workflows/reusable-e2e.yml:238]:
```suggestion
        uses: aws-actions/configure-aws-credentials@e1e17a757e536f70e52b5a12b2e8d1d1c60e04ef # v2.0.0 https://github.com/aws-actions/configure-aws-credentials/commit/e1e17a757e536f70e52b5a12b2e8d1d1c60e04ef
```

It's a general best practice to pin versions in CI.

File: .github/workflows/reusable-e2e.yml [hunk 20]
```diff
@@ -228,6 +234,22 @@ jobs:
...
+        with:
+          aws-region: us-east-2
```
Comment (wlee221) [inline .github/workflows/reusable-e2e.yml:240]:
```suggestion
          aws-region: us-east-1
```

Noticed that this app is in `us-east-1`, not `us-east-2` like other environments. It doesn't have to be this PR since this PR is already large, but we should migrate this environment to `us-east-2` for the sake of consistency.

File: environments/liveness/README.md (hunk 18 above)
Comment (thaddmt) [reply environments/liveness/README.md]: removed!

File: docs/src/components/home/sections/HeroSection.tsx [hunk 21]
```diff
@@ -102,43 +102,37 @@ export const HeroSection = () => {
...
-              className="docs-home-cta"
-              as="a"
-              // @ts-ignore // IGNORE
-              href={`/${platform}/getting-started/installation`}
-            >
-              Get started building
-              <Icon ariaLabel="" as={MdChevronRight} fontSize="xl" />
-            </Button>
-            {platform === 'react' && showEditor ? (
+          {platform === 'swift' || platform === 'android' ? null : (
```
Comment (thaddmt) [inline docs/src/components/home/sections/HeroSection.tsx:109]: @eeatonaws @atierian @reesscot we don't have a `Getting Started` page for android or swift and not sure we'll be able to write one by monday so I just removed those sections on the intro page for now

**Comment by thaddmt** [comment]
Some of the link checker URL's will not work for now as they point to pages that are either private github repos or urls that are still using an allow list like
- https://docs.aws.amazon.com/rekognition/latest/dg/face-liveness-programming-api.html
- https://github.com/aws-amplify/amplify-ui-android/issues/new/choose

@eeatonaws for the amplify-ui-android can we make that package public first thing on 4/10?

File: docs/src/components/home/sections/HeroSection.tsx (hunk 21 above)
Comment (eeatonaws) [reply docs/src/components/home/sections/HeroSection.tsx:109]: I had added a getting started page for Android (page with setup instructions is [here](https://github.com/aws-amplify/amplify-ui/pull/3677/files#diff-fa9b2fad4185c329f46132efd49a17adb89ce8cd7640c701c510fea1fb0f27ba)).

File: examples/next/public/ovalInstruction.png
Comment (hbuchel) [inline examples/next/public/ovalInstruction.png:1]: nit: we can remove this image

File: docs/src/pages/[platform]/connected-components/liveness/customization/customization.theming.swift.mdx
Comment (ioanabrooks) [inline docs/src/pages/[platform]/connected-components/liveness/customization/customization.theming.swift.mdx:1]: Should this file be removed since not implemented?

File: docs/src/pages/_app.page.tsx [hunk 22]
```diff
@@ -55,6 +55,12 @@ if (typeof window === 'undefined') {
 function MyApp({ Component, pageProps }) {
   const [expanded, setExpanded] = React.useState(false);
 
+  Storage.addPluggable(new MyStorageProvider('fast', { delay: 10 }));
+  Storage.addPluggable(new MyStorageProvider('slow', { delay: 1000 }));
+  Storage.addPluggable(
+    new MyStorageProvider('error', { delay: 50, networkError: true })
+  );
+
```
Comment (ioanabrooks) [inline docs/src/pages/_app.page.tsx:63]: Please revert this change as well, pluggables already added on lines 48-53

File: docs/tests/__snapshots__/cssvars-table.test.ts.snap [hunk 23]
```diff
@@ -2538,6 +2538,10 @@ exports[`CSS Variables Table 1`] = `
         \\"variable\\": \\"--amplify-components-link-visited-color\\",
         \\"value\\": \\"var(--amplify-colors-font-interactive)\\"
     },
+    {
+        \\"variable\\": \\"--amplify-components-liveness-camera-module-background-color\\",
+        \\"value\\": \\"var(--amplify-colors-black)\\"
+    },
```
Comment (ioanabrooks) [inline docs/tests/__snapshots__/cssvars-table.test.ts.snap:2544]: Is this the only css var that developers can customize in the entire liveness module?

Comment (thaddmt) [reply docs/tests/__snapshots__/cssvars-table.test.ts.snap:2544]: Hm i'm actually not sure, @hbuchel maybe you can chime in

File: packages/react-liveness/package.json [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    "typecheck": "tsc --noEmit"
+  },
+  "peerDependencies": {
+    "aws-amplify": ">= 5.0.1",
+    "react": ">= 16.14.0",
+    "react-dom": ">= 16.14.0"
+  },
+  "dependencies": {
+    "@aws-amplify/ui": "5.5.7",
+    "@aws-amplify/ui-react": "^4.5.1",
```
Comment (hbuchel) [inline packages/react-liveness/package.json:52]: Question: Should this be pinned to a specific version if we are doing that with /ui?

Comment (thaddmt) [reply packages/react-liveness/package.json:52]: Yep, i'll change it to be pinned

File: docs/tests/__snapshots__/cssvars-table.test.ts.snap (hunk 23 above)
Comment (hbuchel) [reply docs/tests/__snapshots__/cssvars-table.test.ts.snap:2544]: Yeeeeeeahh, I think this was the start of us moving some of the inline css to tokens/classes last year but I'm not sure why we stopped there. We should definitely follow up with some more cleanup there so the tokens can be themed separately (i.e. so the border around the camera view can be themed separately from .colors.border.tertiary, etc.)

**Review by hbuchel — APPROVED** [review]
🚢

File: docs/src/pages/[platform]/connected-components/liveness/customization/customization.theming.swift.mdx
Comment (atierian) [reply docs/src/pages/[platform]/connected-components/liveness/customization/customization.theming.swift.mdx:1]: Addressed in [351ad95](https://github.com/aws-amplify/amplify-ui/pull/3677/commits/351ad9562701a0ff983fe8da13930e483792f66c)

File: docs/tests/__snapshots__/cssvars-table.test.ts.snap (hunk 23 above)
Comment (ioanabrooks) [reply docs/tests/__snapshots__/cssvars-table.test.ts.snap:2544]: Thanks for confirming. We should have a follow-up ticket for this.

**Review by wlee221 — COMMENTED** [review]
Since link checker is expected to fail, let's set [`continue-on-error`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error) to true on that step, so that CI doesn't get blocked when we try to publish to `@next`.

File: canary/e2e/features/liveness/face-detect.feature (hunk 5 above)
Comment (ioanabrooks) [reply canary/e2e/features/liveness/face-detect.feature:6]: We removed the Liveness component from the example so this test will fail. Add `@skip` tag or remove this file or this test will fail.

**Review by ioanabrooks — DISMISSED** [review]

**Review by hbuchel — DISMISSED** [review]

**Review by wlee221 — DISMISSED** [review]
LGTM. After merge, we need to:

1. Reenable migrate checker once all the links are available
2. Add back liveness test to build canaries
3. Migrate liveness environment to us-east-2. 

(in order of high priority to low priority).

**Review by reesscot — APPROVED** [review]

**Review by wlee221 — APPROVED** [review]

**Review by ioanabrooks — APPROVED** [review]
